File: library/src/main/java/com/firemaples/googlewebtranslator/GoogleWebTranslator.java
package com.firemaples.googlewebtranslator;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.net.http.SslError;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;
import android.webkit.SslErrorHandler;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.loopj.android.http.SyncHttpClient;
import com.loopj.android.http.TextHttpResponseHandler;

import java.io.ByteArrayInputStream;
import java.io.UnsupportedEncodingException;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import cz.msebera.android.httpclient.Header;

@SuppressWarnings({"FieldCanBeLocal", "unused"})
public class GoogleWebTranslator {
    private static final String TAG = GoogleWebTranslator.class.getSimpleName();

    private final static String FORMAT_TEXT = "{TEXT}";
    //    private final static String FORMAT_SOURCE_LANG = "{SL}";
    private final static String FORMAT_TARGET_LANG = "{TL}";

    private static String URL_GOOGLE_TRANSLATE = "https://translate.google.com/m/translate?sl=auto&tl={TL}&ie=UTF-8";
    private static String URL_LOAD_TRANSLATION_RESULT = "https://translate.google.com/translate_a/single?";

    private static String JS_FORMAT = "javascript:%s;void 0";
    private static String JS_INIT = "document.getElementsByClassName('translation')[0].addEventListener('DOMNodeInserted', function(e){window.HtmlViewer.onTranslationSuccess(e.target.textContent);})";
    private static String JS_TRANSLATE = "document.getElementById('source').value = '{TEXT}'";
//    private static String JS_TRANSLATION_FAILED_CHECK = "if(document.getElementsByClassName('translation-error')[0].style.display === ''){onTranslationFailed();}";
//    private static String JS_FORCE_POST_TRANSLATE = "var text='%s';for(var i=0;i<700;i++){text+=i};document.getElementById('source').value = text";

//    private static String POST_FORM_Q = "q";

    private static String HTTP_GET = "GET";
    private static String UTF_8 = "UTF-8";


    @SuppressLint("StaticFieldLeak")
    private static GoogleWebTranslator _instance = null;

    private WeakReference<Context> weakContext;

    private WebView webView;
    private String currentTranslatorUrl;
    private boolean receivedErrorBeforeInitialized = false;
    private boolean initialized = false;
    private String textToTranslate = null;
    private long prepareStartTime = 0, translationStartTime = 0;

    private boolean callbackOnMainThread = true;
    @VisibleForTesting
    protected List<OnTranslationCallback> callbackList = new ArrayList<>();
    private Handler backgroundHandler = new Handler();

    public static synchronized GoogleWebTranslator init(Context context) {
        if (_instance == null) {
            _instance = new GoogleWebTranslator(context);
        }
        return _instance;
    }

    public static synchronized GoogleWebTranslator getInstance() {
        if (_instance == null) {
            throw new IllegalStateException("Please initialize GoogleWebTranslator be for using.");
        }
        return _instance;
    }

    private GoogleWebTranslator(Context context) {
        this.weakContext = new WeakReference<>(context);
        _init();
    }

    private Context getContext() {
        if (weakContext == null) {
            throw new IllegalStateException("Please initialize GoogleWebTranslator be for using.");
        } else {
            Context context = weakContext.get();
            if (context == null) {
                throw new NullPointerException("Context is null");
            } else {
                return context;
            }
        }
    }

    public WebView getNonParentWebView() {
        ViewGroup viewGroup = (ViewGroup) webView.getParent();
        if (viewGroup != null) {
            viewGroup.removeView(webView);
        }

        return webView;
    }

    @SuppressLint({"ClickableViewAccessibility", "SetJavaScriptEnabled"})
    private void _init() {
        webView = new WebView(getContext());
        webView.addJavascriptInterface(new MyJavaScriptInterface(), "HtmlViewer");
        webView.setWebViewClient(new MyWebViewClient());
        webView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                return false;
            }
        });
        webView.setLayerType(View.LAYER_TYPE_HARDWARE, null);
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);

//        settings.setAppCacheEnabled(false);
        settings.setBuiltInZoomControls(false);
        settings.setBlockNetworkImage(true);
        settings.setDatabaseEnabled(false);
        settings.setGeolocationEnabled(false);
        settings.setJavaScriptCanOpenWindowsAutomatically(false);
        settings.setLoadsImagesAutomatically(false);
        settings.setSaveFormData(false);
        //noinspection deprecation
        settings.setSavePassword(false);
        settings.setSupportMultipleWindows(false);
    }

    public void addOnTranslationCallback(OnTranslationCallback onTranslationCallback) {
        this.callbackList.add(onTranslationCallback);
    }

    public void removeOnTranslationCallback(OnTranslationCallback onTranslationCallback) {
        this.callbackList.remove(onTranslationCallback);
    }

    public void clearAllCallbacks() {
        this.callbackList.clear();
    }

    public void setCallbackOnMainThread(boolean callbackOnMainThread) {
        this.callbackOnMainThread = callbackOnMainThread;
    }

    public void setTargetLanguage(Language targetLanguage) {
        setTargetLanguage(targetLanguage.getLangCode());
    }

    @SuppressWarnings("WeakerAccess")
    public void setTargetLanguage(String targetLanguage) {
        String url = URL_GOOGLE_TRANSLATE.replace(FORMAT_TARGET_LANG, targetLanguage);

        LogTool.i(TAG, "setTargetLanguage: " + url);

        currentTranslatorUrl = url;

        _loadTranslatorUrl();
    }

    private void _loadTranslatorUrl() {
        prepareStartTime = System.currentTimeMillis();
        receivedErrorBeforeInitialized = false;
        initialized = false;

        runOnMainThread(new Runnable() {
            @Override
            public void run() {
                webView.loadUrl(currentTranslatorUrl);
            }
        });
    }

    private void runOnMainThread(Runnable runnable) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            runnable.run();
        } else {
            webView.post(runnable);
        }
    }

    public void translate(String text) {
        textToTranslate = text;

        if (!initialized) {
            _loadTranslatorUrl();
        } else {
//        text = text.replaceAll("\\R", "\\n");
//        String textToSend = String.format(JS_FORCE_POST_TRANSLATE, String.valueOf(translationCounter = !translationCounter));
//        _doJavascript(textToSend);

            translationStartTime = System.currentTimeMillis();

            _doJavascript(JS_TRANSLATE.replace(FORMAT_TEXT, text));
        }
    }

    private void _doJavascript(String javascript) {
        final String url = String.format(Locale.US, JS_FORMAT, javascript);

        LogTool.i(TAG, "_doJavascript: " + url);

        runOnMainThread(new Runnable() {
            @Override
            public void run() {
                webView.loadUrl(url);
            }
        });
    }

    private void _postCallbackToThread(Runnable runnable) {
        if (callbackOnMainThread) {
            webView.post(runnable);
        } else {
            backgroundHandler.post(runnable);
        }
    }

    private void postInitialized() {
        for (final OnTranslationCallback callback : callbackList) {
            _postCallbackToThread(new Runnable() {
                @Override
                public void run() {
                    callback.onInitialized(GoogleWebTranslator.this);
                }
            });
        }
    }

    private void postInitializationFailed(final int errorCode, final String desc) {
        for (final OnTranslationCallback callback : callbackList) {
            _postCallbackToThread(new Runnable() {
                @Override
                public void run() {
                    callback.onInitializationFailed(GoogleWebTranslator.this, errorCode, desc);
                }
            });
        }
    }

    private void postTranslationFinish(final TranslatedResult result) {
        for (final OnTranslationCallback callback : callbackList) {
            _postCallbackToThread(new Runnable() {
                @Override
                public void run() {
                    callback.onTranslationSuccess(GoogleWebTranslator.this, result);
                }
            });
        }
    }

    private void postTranslationFailed(final Throwable throwable) {
        for (final OnTranslationCallback callback : callbackList) {
            _postCallbackToThread(new Runnable() {
                @Override
                public void run() {
                    callback.onTranslationFailed(GoogleWebTranslator.this, throwable);
                }
            });
        }
    }

    private class MyWebViewClient extends WebViewClient {
        private SyncHttpClient httpClient = new SyncHttpClient();

        private MyWebViewClient() {
            httpClient.setCookieStore(CookieStoreUtil.getCookieStore());
        }

        @Override
        public void onPageFinished(WebView view, String url) {
            super.onPageFinished(view, url);

            if (!receivedErrorBeforeInitialized) {
                _doJavascript(JS_INIT);
                initialized = true;
                long preparedSpentTime = System.currentTimeMillis() - prepareStartTime;
                LogTool.i(TAG, "Page initialized, spent: " + preparedSpentTime + " ms");

                postInitialized();

                if (textToTranslate != null) {
                    translate(textToTranslate);
                }
            }
        }

        @Override
        public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
            super.onReceivedError(view, errorCode, description, failingUrl);
            if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                return;
            }
            LogTool.e(TAG, "onReceivedError(), errorCode: " + errorCode + ", desc: " + description);

            _onReceivedError(failingUrl, errorCode, description);
        }

        @TargetApi(android.os.Build.VERSION_CODES.M)
        @Override
        public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
            super.onReceivedError(view, request, error);
            LogTool.e(TAG, "onReceivedError()");

            _onReceivedError(request.getUrl().toString(), error.getErrorCode(), error.getDescription().toString());
        }

        private void _onReceivedError(String url, int errorCode, String desc) {
            if (!initialized && url.equalsIgnoreCase(currentTranslatorUrl)) {
                receivedErrorBeforeInitialized = true;
                initialized = false;
                postInitializationFailed(errorCode, desc);
            }
        }

        @Override
        public void onReceivedHttpError(WebView view, WebResourceRequest request, WebResourceResponse errorResponse) {
            super.onReceivedHttpError(view, request, errorResponse);
        }

        @Override
        public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
            super.onReceivedSslError(view, handler, error);
        }

        // shouldInterceptRequest:
        // https://gist.github.com/kibotu/32313b957cd01258cf67

        @Nullable
        @Override
        public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
            String url = request.getUrl().toString();
            boolean methodGet = HTTP_GET.equalsIgnoreCase(request.getMethod());
            Map<String, String> requestHeaders = request.getRequestHeaders();

            WebResourceResponse response = _interceptRequest(url, methodGet, requestHeaders);
            if (response != null) {
                return response;
            } else {
                return super.shouldInterceptRequest(view, request);
            }
        }

//        @Nullable
//        @Override
//        public WebResourceResponse shouldInterceptRequest(WebView view, String url) {
//
////            WebResourceResponse response = _interceptRequest(url, true, null);
////
////            if (response != null) {
////                return response;
////            } else {
//            return super.shouldInterceptRequest(view, url);
////            }
//        }

        private long getTranslationSpentTime() {
            return System.currentTimeMillis() - translationStartTime;
        }

        private WebResourceResponse _interceptRequest(String url, boolean methodGet, Map<String, String> requestHeaders) {
            if (!url.startsWith(URL_LOAD_TRANSLATION_RESULT)) {
                return null;
            }

            LogTool.i(TAG, "_interceptRequest(), url: " + url
                    + ", methodGet: " + methodGet
                    + ", requestHeaders!=null: " + String.valueOf(requestHeaders != null));

            if (!methodGet) {
                LogTool.e(TAG, "POST not working now, spent: " + getTranslationSpentTime() + " ms");

                postTranslationFailed(new IllegalArgumentException("Http POST is not supported now"));

                return null;
            }

            final boolean[] success = {false};
            final Header[][] responseHeaders = new Header[1][1];
            final String[] responseText = new String[1];
            final Throwable[] httpError = new Throwable[1];

//            if (methodGet) {
//                try {
//                    url = removeQueryParameter(url, POST_FORM_Q);
//                } catch (URISyntaxException e) {
//                    e.printStackTrace();
//                }
//            }

            //Convert headers
            Header[] headersToSend = HttpUtil.getHeaders(requestHeaders);

            //Convert cookie
            CookieStoreUtil.syncFromWebView(url);

//            if (methodGet) {
            httpClient.get(getContext(), url, headersToSend, null, new TextHttpResponseHandler(UTF_8) {
                @Override
                public void onFailure(int statusCode, Header[] headers, String responseString, Throwable throwable) {
                    success[0] = false;
                    httpError[0] = throwable;
                }

                @Override
                public void onSuccess(int statusCode, Header[] headers, String responseString) {
                    success[0] = true;
                    responseHeaders[0] = headers;
                    responseText[0] = responseString;
                }
            });
//            } else {
//                //TODO not working now
//
//                Map<String, String> formData = new HashMap<>();
//                formData.put(POST_FORM_Q, textToTranslate);
////            formData.putAll(requestHeaders);
//
//                StringEntity stringEntity = null;
//                try {
//                    stringEntity = new StringEntity(POST_FORM_Q + "=" + textToTranslate);
//                } catch (UnsupportedEncodingException e) {
//                    e.printStackTrace();
//                }
//
//                httpClient.post(getContext(), url, headersToSend, stringEntity, "application/json; charset=UTF-8", new TextHttpResponseHandler(UTF_8) {
//                    @Override
//                    public void onFailure(int statusCode, Header[] headers, String responseString, Throwable throwable) {
//                        success[0] = false;
//                        httpError[0] = throwable;
//                    }
//
//                    @Override
//                    public void onSuccess(int statusCode, Header[] headers, String responseString) {
//                        success[0] = true;
//                        responseHeaders[0] = headers;
//                        responseText[0] = responseString;
//                    }
//                });
//            }

            if (success[0]) {
                String raw = responseText[0];

                TranslatedResult result = ResultParser.parse(raw);

                LogTool.i(TAG, "On result success, spent: " + getTranslationSpentTime() + " ms, result: " + raw);

                postTranslationFinish(result);

                try {
                    return new WebResourceResponse(
                            null,
                            UTF_8,
                            new ByteArrayInputStream(raw.getBytes(UTF_8))
                    ) {
                        @Override
                        public Map<String, String> getResponseHeaders() {
                            Map<String, String> headers = new HashMap<>();
                            for (Header header : responseHeaders[0]) {
                                headers.put(header.getName(), header.getValue());
                            }
                            return headers;
                        }
                    };
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                    return null;
                }
            } else {
                LogTool.e(TAG, "Result failed: " + httpError[0].getMessage());

                postTranslationFailed(httpError[0]);

                return null;
            }
        }

        @Override
        public void onLoadResource(WebView view, String url) {
            super.onLoadResource(view, url);

            LogTool.d(TAG, "onLoadResource: " + url);
        }
    }

    @SuppressWarnings("unused")
    private class MyJavaScriptInterface {
        @JavascriptInterface
        public void onTranslationSuccess(String text) {
            LogTool.d(TAG, "onTranslationSuccess: " + text);
        }

        @JavascriptInterface
        public void onTranslationFailed() {
            LogTool.d(TAG, "onTranslationFailed");
        }
    }

    public interface OnTranslationCallback {
        void onInitialized(GoogleWebTranslator translator);

        void onInitializationFailed(GoogleWebTranslator translator, int errorCode, String desc);

        void onTranslationSuccess(GoogleWebTranslator translator, TranslatedResult result);

        void onTranslationFailed(GoogleWebTranslator translator, Throwable throwable);
    }
}

File: library/src/main/java/com/firemaples/googlewebtranslator/GoogleWebTranslator.kt
package com.firemaples.googlewebtranslator

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.support.annotation.RequiresApi
import android.util.Log
import android.view.ViewGroup
import android.webkit.*
import kotlinx.coroutines.Job
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.ByteArrayInputStream

@SuppressLint("SetJavaScriptEnabled")
class GoogleWebTranslator(private val context: Context) {
    companion object {
        internal val tag: String = GoogleWebTranslator::class.java.simpleName
        private var stethoInitialized = false
    }

    private val maxTextSize = 5000

    private val webView: WebView by lazy {
        WebView(context).also {
            it.webViewClient = MyWebViewClient(context)
//            setLayerType(View.LAYER_TYPE_HARDWARE, null)

            it.addJavascriptInterface(MyJSInterface(it), "MyJS")

            it.settings.apply {
                javaScriptEnabled = true
//                builtInZoomControls = false
                blockNetworkImage = true
                databaseEnabled = false
                setGeolocationEnabled(false)
                javaScriptCanOpenWindowsAutomatically = false
                loadsImagesAutomatically = false
                setSupportMultipleWindows(false)
//                setSupportZoom(false)
            }
        }
    }

    public fun setup(parentView: ViewGroup) {
//        if (DEBUG && !stethoInitialized) {
//            Stetho.initializeWithDefaults(context)
//        }
        threadUI.launch {
            webView.parent?.also { (it as ViewGroup).removeView(webView) }
            parentView.addView(webView,
                    ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        }
    }

    public fun translate(_text: String, targetLang: String, callback: OnTranslationCallback) {
        threadUI.launch {
            if (webView.parent == null) {
                LogTool.e(tag, "Please call setup() before translation", true)
                return@launch
            }

            val text = _text.let {
                if (it.length > maxTextSize) {
                    LogTool.w(tag, "The maximum request text is limited to $maxTextSize, current is ${it.length}", true)
                    it.substring(0, maxTextSize)
                } else it
            }

            LogTool.d(tag, "Text to request[${text.length}]: $text")
            val url = "https://translate.google.com/?sl=auto&tl=$targetLang&ie=UTF-8&op=translate"
            webView.setTag(R.id.text, text)

            LogTool.i(tag, "load url: $url")
            webView.setTag(R.id.url, url)
            webView.setTag(R.id.startTime, System.currentTimeMillis())
            webView.setTag(R.id.callback, callback)

            threadUI.launch {
                callback.onStart()
            }

            webView.loadUrl(url)
        }
    }

    private class MyWebViewClient(val context: Context) : WebViewClient() {
        private val httpGet = "GET"
        private val utf8 = "UTF-8"
        private val urlLoadTranslationResult = "https://translate.google.com/translate_a/single?"
        private val tempABCD = "tempABCD"

        private val htmlTitleParser by lazy { Regex("<title>([\\s\\S]*?)<\\/title>") }


        val interceptorJS: String by lazy {
            context.assets.open("intercept-network-requests.js").bufferedReader().use { it.readText() }
        }

        private val okHttpClient by lazy {
            OkHttpClient.Builder().apply {
                cookieJar(WebViewCookieHandler())
//                if (DEBUG) {
//                    addNetworkInterceptor(StethoInterceptor())
//                }
            }.build()
        }

        override fun onPageFinished(view: WebView, url: String) {
            super.onPageFinished(view, url)
            LogTool.i(tag, "onPageFinished, url: $url")

            val tagText = view.getTag(R.id.text)
            if (tagText != null) {
                view.setTag(R.id.text, null)

                val text = tagText.toString()

                text.chunked(100).forEachIndexed { index, s ->
                    val js = if (index == 0) {
                        "$tempABCD=\"$s\""
                    } else {
                        "$tempABCD+=\"$s\""
                    }
                    loadJS(view, js)
                }

                val sourceTextArea = "document.getElementsByTagName('textarea')[0]"
                loadJS(view, "$sourceTextArea.value=$tempABCD")
                loadJS(view, "function eventFire(el, etype){\n" +
                        "  if (el.fireEvent) {\n" +
                        "    el.fireEvent('on' + etype);\n" +
                        "  } else {\n" +
                        "    var evObj = document.createEvent('Events');\n" +
                        "    evObj.initEvent(etype, true, false);\n" +
                        "    el.dispatchEvent(evObj);\n" +
                        "  }\n" +
                        "}")
                loadJS(view, interceptorJS)
                loadJS(view, "$sourceTextArea.dispatchEvent(new Event('input', { bubbles: true}))")

                loadJSAsync(view, "$sourceTextArea.value") {
                    val requestText = it.substring(1, it.length - 1)
                    LogTool.i(tag, "Got 'source' value[${requestText.length}]: $requestText")
                    if (!it.isBlank()) {
                        view.setTag(R.id.textPost, requestText)
//                        val translateButton = "document.querySelectorAll(\"[aria-label='Translate']\")[0]"
//                        loadJS(view, "eventFire($translateButton,'mousedown')")
                        loadJS(view, "$sourceTextArea.dispatchEvent(new Event('input', { bubbles: true}))")
                    }
                }
            }
        }

        override fun onReceivedError(view: WebView, errorCode: Int, description: String?, failingUrl: String?) {
            super.onReceivedError(view, errorCode, description, failingUrl)
            handleError(view, errorCode, description, failingUrl)
        }

        @RequiresApi(Build.VERSION_CODES.M)
        override fun onReceivedError(view: WebView, request: WebResourceRequest?, error: WebResourceError?) {
            super.onReceivedError(view, request, error)
            handleError(view, error?.errorCode, error?.description?.toString(), request?.url.toString())
        }

        private fun handleError(view: WebView, errorCode: Int?, description: String?, failingUrl: String?) {
            LogTool.e(tag, "handleError(), url: $failingUrl, error: ($errorCode)$description")
            val originUrl = view.getTag(R.id.url) as String?
            if (originUrl != null && failingUrl == originUrl) {
                view.postCallback {
                    it.onTranslationFailed("($errorCode) $description")
                }
            }
        }

        private fun loadJS(view: WebView, js: String) {
            view.postDelayed({
                val fullJS = "javascript:$js;void 0"
                LogTool.i(tag, "load JS: $fullJS")
                view.loadUrl(fullJS)
            }, 0)
        }

        private fun loadJSAsync(view: WebView, js: String, callback: ((String) -> Unit)? = null) {
            view.postDelayed({
                val fullJS = "javascript:$js"
                LogTool.i(tag, "load JSAsync: $fullJS")
                view.evaluateJavascript(fullJS) {
                    callback?.invoke(it)
                }
            }, 0)
        }

//        override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? {
//            LogTool.d(tag, "shouldInterceptRequest: ${request.url}")
//            if (request.url.toString().startsWith(urlLoadTranslationResult, true)) {
//                val response = doInterceptRequest(view, request)
//                if (response != null) {
//                    return response
//                }
//            }
//            return super.shouldInterceptRequest(view, request)
//        }
//
//        private fun doInterceptRequest(view: WebView, originReq: WebResourceRequest): WebResourceResponse? {
//            LogTool.i(tag, "doInterceptRequest: ${originReq.method} ${originReq.url}")
//
//            val request = Request.Builder().apply {
//                url(originReq.url.toString())
//                when {
//                    httpGet.equals(originReq.method, true) ->
//                        method(httpGet, null)
//                    else ->
//                        method(originReq.method,
//                                FormBody.Builder().add("q",
//                                        view.getTag(R.id.textPost).toString()).build())
//                }
//                originReq.requestHeaders.forEach {
//                    header(it.key, it.value)
//                }
//            }.build()
//
//            LogTool.i(tag, "intercept: $request")
//
//            var errorMsg: String? = null
//
//            try {
//                val response = okHttpClient.newCall(request).execute()
//                val headers = response.headers().toMultimap()
//                        .map { it.key to it.value.firstOrNull() }.toMap()
//                val status = response.code()
//                val body = response.body()?.string()
//                val inputStream = body?.let {
//                    ByteArrayInputStream(it.toByteArray())
//                }
//                val timeSpent = (view.getTag(R.id.startTime) as Long?)?.let {
//                    "${System.currentTimeMillis() - it}ms"
//                }
//
//                if (response.isSuccessful) {
//                    if (body != null) {
//                        LogTool.i(tag, body)
//                        val result = ResultParser.parse(body)
//                        LogTool.i(tag, "result($timeSpent): ${result.text}")
//                        view.postCallback {
//                            it.onTranslated(result)
//                        }
//                        return WebResourceResponse(null, utf8,
//                                status, "OK",
//                                headers,
//                                inputStream)
//                    } else {
//                        LogTool.w(tag, "Response body is null($timeSpent)")
////                        return WebResourceResponse(null, utf8,
////                                status, "OK",
////                                headers,
////                                inputStream)
//                    }
//                } else {
//                    errorMsg = if (body != null) {
//                        htmlTitleParser.find(body)?.groupValues?.let {
//                            if (it.size >= 2) {
//                                it[1]
//                            } else null
//                        }
//                    } else null
//                    LogTool.w(tag, "Response failed($timeSpent): $status, $errorMsg, $body")
////                    return WebResourceResponse(null, utf8, status, "OK", headers, inputStream)
//                }
//            } catch (t: Throwable) {
//                LogTool.e(tag, "Error shown: ${Log.getStackTraceString(t)}")
//                errorMsg = t.localizedMessage
//            }
//
//            view.postCallback {
//                it.onTranslationFailed(errorMsg ?: "Unexpected result")
//            }
//
//            return null
//        }
    }

    @Suppress("unused")
    private class MyJSInterface(val webView: WebView) {
        @JavascriptInterface
        fun onTranslatedResult(text: String) {
            LogTool.d(tag, "On translated, $text")

            webView.postCallback {
                it.onTranslated(TranslatedResult().apply { this.text = text })
            }
        }
    }

    interface OnTranslationCallback {
        fun onStart()
        fun onTranslated(result: TranslatedResult)
        fun onTranslationFailed(errorMsg: String)
    }
}

private fun WebView.postCallback(block: (GoogleWebTranslator.OnTranslationCallback) -> Unit): Job = threadUI.launch {
    this@postCallback.getTag(R.id.callback)?.also {
        val callback = it as GoogleWebTranslator.OnTranslationCallback
        this@postCallback.setTag(R.id.callback, null)

        block(callback)
    }
}